pkg/autoscaler/autoscaler.go

```diff
@@ -674,4 +675,96 @@ var _ = Describe("[Feature:Machines] Autoscaler should", func() {
 			}
 		})
 	})
+
+	Context("use a ClusterAutoscaler that has 12 maximum total nodes count and balance similar nodes enabled", func() {
+		var clusterAutoscaler *caov1.ClusterAutoscaler
+
+		BeforeEach(func() {
+			By("Creating ClusterAutoscaler")
+			clusterAutoscaler = clusterAutoscalerResource(12)
+			clusterAutoscaler.Spec.BalanceSimilarNodeGroups = pointer.BoolPtr(true)
+			Expect(client.Create(ctx, clusterAutoscaler)).Should(Succeed())
+			cleanupObjects[clusterAutoscaler.GetName()] = clusterAutoscaler
+		})
+
+		AfterEach(func() {
+			// explicitly delete the ClusterAutoscaler
+			// this is needed due to the autoscaler tests requiring singleton
+			// deployments of the ClusterAutoscaler.
+			By("Waiting for ClusterAutoscaler to delete.")
+			caName := clusterAutoscaler.GetName()
+			Expect(deleteObject(caName, cleanupObjects[caName])).Should(Succeed())
+			delete(cleanupObjects, caName)
+			Eventually(func() bool {
+				t := &apierrors.StatusError{}
+				// Convert the error to a StatusError and allow `IsNotFound` to check that.
+				// TODO drop this conversion once we have upgraded to K8s 1.19 which will support error wrapping first class.
+				if _, err := framework.GetClusterAutoscaler(client, caName); err != nil && errors.As(err, &t) && apierrors.IsNotFound(t) {
+					return true
+				}
+				return false
+			}, framework.WaitMedium, pollingInterval).Should(BeTrue())
+		})
+
+		It("places nodes evenly across node groups [Slow]", func() {
+			By("Creating 2 MachineSets each with 1 replica")
+			var transientMachineSets [2]*mapiv1beta1.MachineSet
+			targetedNodeLabel := fmt.Sprintf("%v-delete-cleanup", autoscalerWorkerNodeRoleLabel)
+			for i, machineSet := range transientMachineSets {
+				machineSetParams := framework.BuildMachineSetParams(client, 1)
+				machineSetParams.Labels[targetedNodeLabel] = ""
+				// remove this label to make the MachineSets similar, see below for more details
+				delete(machineSetParams.Labels, "e2e.openshift.io")
+				machineSet, err = framework.CreateMachineSet(client, machineSetParams)
+				Expect(err).ToNot(HaveOccurred())
+				cleanupObjects[machineSet.GetName()] = machineSet
+				transientMachineSets[i] = machineSet
+			}
+
+			// balance similar nodes requires that all the participating MachineSets have the same
+			// instance types and labels. while the instance types should be the same, we want to
```

Comment:
Balance similar nodes should work off of CPU/RAM/GPU provided primarily, then instance types as a fall back if that info is not available.

Why do the labels need to be the same?  Not saying your tests are wrong, but that seems like a broken implementation.  Or is this magic all just needed for the tests?

Reply:
this is a limitation coming from the autoscaler core, [referenced here](https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/FAQ.md#im-running-cluster-with-nodes-in-multiple-zones-for-ha-purposes-is-that-supported-by-cluster-autoscaler)

> If you set the flag to true, CA will automatically identify node groups with the same instance type and the same set of labels (except for automatically added zone label) and try to keep the sizes of those node groups balanced.

iirc, MachineSets basically equate to node groups internally for the autoscaler, so it's just asking for all the labels and then checking them (at the autoscaler). this might be inefficient behavior from the autoscaler, but thinking about normalizing the labels before returning to the autoscaler seems like it could cause more problems.